File: PrometheusToKomodor/webhook-middleware.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"log/slog"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	defaultListenAddr     = ":8080"
	defaultKomodorURL     = "https://api.komodor.com/mgmt/v1/events"
	defaultRequestTimeout = 10 * time.Second
)

type AlertmanagerPayload struct {
	Receiver string              `json:"receiver"`
	Status   string              `json:"status"`
	Alerts   []AlertmanagerAlert `json:"alerts"`
	GroupKey string              `json:"groupKey"`
}

type AlertmanagerAlert struct {
	Status       string            `json:"status"`
	Labels       map[string]string `json:"labels"`
	Annotations  map[string]string `json:"annotations"`
	StartsAt     string            `json:"startsAt"`
	EndsAt       string            `json:"endsAt"`
	GeneratorURL string            `json:"generatorURL"`
	Fingerprint  string            `json:"fingerprint"`
}

type KomodorEvent struct {
	EventType string            `json:"eventType"`
	Summary   string            `json:"summary"`
	Scope     KomodorScope      `json:"scope"`
	Severity  string            `json:"severity,omitempty"`
	Details   map[string]string `json:"details,omitempty"`
}

type KomodorScope struct {
	Clusters      []string `json:"clusters"`
	Namespaces    []string `json:"namespaces,omitempty"`
	ServicesNames []string `json:"servicesNames,omitempty"`
}

type Server struct {
	logger     *slog.Logger
	client     *http.Client
	komodorURL string
	komodorKey string
}

func main() {
	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))

	server := &Server{
		logger: logger,
		client: &http.Client{
			Timeout: envDuration("REQUEST_TIMEOUT", defaultRequestTimeout),
		},
		komodorURL: envString("KOMODOR_API_URL", defaultKomodorURL),
		komodorKey: os.Getenv("KOMODOR_API_KEY"),
	}

	if server.komodorKey == "" {
		logger.Error("KOMODOR_API_KEY is not set, exiting")
		os.Exit(1)
	}

	mux := http.NewServeMux()
	mux.HandleFunc("/healthz", server.healthzHandler)
	mux.HandleFunc("/alertmanager", server.alertmanagerHandler)

	listenAddr := envString("LISTEN_ADDR", defaultListenAddr)

	logger.Info("starting alertmanager komodor adapter", "listen_addr", listenAddr)

	if err := http.ListenAndServe(listenAddr, mux); err != nil {
		logger.Error("server stopped", "error", err)
		os.Exit(1)
	}
}

func (s *Server) healthzHandler(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

func (s *Server) alertmanagerHandler(w http.ResponseWriter, r *http.Request) {
	if s.komodorKey == "" {
		http.Error(w, "KOMODOR_API_KEY is not configured", http.StatusInternalServerError)
		return
	}

	var payload AlertmanagerPayload
	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
		http.Error(w, "invalid alertmanager payload", http.StatusBadRequest)
		return
	}

	sent := 0

	for _, alert := range payload.Alerts {
		event, err := buildKomodorEvent(alert)
		if err != nil {
			s.logger.Warn(
				"skipping alert without required cluster label",
				"alertname", alert.Labels["alertname"],
				"fingerprint", alert.Fingerprint,
			)
			continue
		}

		if err := s.sendKomodorEvent(r.Context(), event); err != nil {
			s.logger.Error(
				"failed to send komodor event",
				"alertname", alert.Labels["alertname"],
				"fingerprint", alert.Fingerprint,
				"error", err,
			)

			http.Error(w, "failed to send komodor event", http.StatusBadGateway)
			return
		}

		sent++
	}

	writeJSON(w, http.StatusOK, map[string]int{
		"received": len(payload.Alerts),
		"sent":     sent,
	})
}

func buildKomodorEvent(alert AlertmanagerAlert) (KomodorEvent, error) {
	cluster := strings.TrimSpace(alert.Labels["cluster"])
	if cluster == "" {
		return KomodorEvent{}, errors.New("missing required cluster label")
	}

	alertName := valueOrDefault(alert.Labels["alertname"], "AlertmanagerAlert")
	eventType := truncate(alertName, 30)

	summary := firstNonEmpty(
		alert.Annotations["summary"],
		alert.Annotations["description"],
		strings.ToTitle(alert.Status)+": "+alertName,
	)

	namespace := firstNonEmpty(
		alert.Labels["namespace"],
		alert.Labels["kubernetes_namespace"],
	)

	service := firstNonEmpty(
		alert.Labels["service"],
		alert.Labels["service_name"],
		alert.Labels["app"],
		alert.Labels["app_kubernetes_io_name"],
		alert.Labels["deployment"],
		alert.Labels["statefulset"],
		alert.Labels["daemonset"],
	)

	details := map[string]string{
		"status":       alert.Status,
		"startsAt":     alert.StartsAt,
		"endsAt":       alert.EndsAt,
		"generatorURL": alert.GeneratorURL,
		"fingerprint":  alert.Fingerprint,
	}

	for key, value := range alert.Labels {
		details["label_"+key] = value
	}

	for key, value := range alert.Annotations {
		details["annotation_"+key] = value
	}

	return KomodorEvent{
		EventType: eventType,
		Summary:   summary,
		Scope: KomodorScope{
			Clusters:      []string{cluster},
			Namespaces:    optionalSlice(namespace),
			ServicesNames: optionalSlice(service),
		},
		Severity: mapSeverity(alert.Labels["severity"], alert.Status),
		Details:  details,
	}, nil
}

func (s *Server) sendKomodorEvent(ctx context.Context, event KomodorEvent) error {
	body, err := json.Marshal(event)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(
		ctx,
		http.MethodPost,
		s.komodorURL,
		bytes.NewReader(body),
	)
	if err != nil {
		return err
	}

	req.Header.Set("content-type", "application/json")
	req.Header.Set("x-api-key", s.komodorKey)

	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}
	// Ensure the response body is closed to prevent resource leaks
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusCreated {
		responseBody, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))

		return errors.New("komodor api returned " + resp.Status + ": " + string(responseBody))
	}

	return nil
}

func mapSeverity(severity string, status string) string {
	if status == "resolved" {
		return "information"
	}

	switch strings.ToLower(severity) {
	case "critical", "error", "page":
		return "error"
	case "warning", "warn":
		return "warning"
	default:
		return "information"
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}

	return ""
}

func valueOrDefault(value string, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}

	return value
}

func optionalSlice(value string) []string {
	if strings.TrimSpace(value) == "" {
		return nil
	}

	return []string{value}
}

func truncate(value string, maxLength int) string {
	if len(value) <= maxLength {
		return value
	}

	return value[:maxLength]
}

func envString(key string, fallback string) string {
	value := os.Getenv(key)
	if value == "" {
		return fallback
	}

	return value
}

func envDuration(key string, fallback time.Duration) time.Duration {
	value := os.Getenv(key)
	if value == "" {
		return fallback
	}

	duration, err := time.ParseDuration(value)
	if err != nil {
		return fallback
	}

	return duration
}

func writeJSON(w http.ResponseWriter, statusCode int, body any) {
	w.Header().Set("content-type", "application/json")
	w.WriteHeader(statusCode)

	_ = json.NewEncoder(w).Encode(body)
}

File: PrometheusToKomodor/python/webhook-middleware.py
from __future__ import annotations

import os
from typing import Any

import httpx
from fastapi import FastAPI, HTTPException, Request
from pydantic import BaseModel, Field


KOMODOR_API_URL = os.getenv(
    "KOMODOR_API_URL",
    "https://api.komodor.com/mgmt/v1/events",
)
KOMODOR_API_KEY = os.getenv("KOMODOR_API_KEY", "")
REQUEST_TIMEOUT_SECONDS = float(os.getenv("REQUEST_TIMEOUT_SECONDS", "10"))

app = FastAPI(title="Alertmanager Komodor Adapter")


class KomodorScope(BaseModel):
    clusters: list[str]
    namespaces: list[str] | None = None
    servicesNames: list[str] | None = None  # Komodor schema spelling


class KomodorEvent(BaseModel):
    eventType: str = Field(..., max_length=30)
    summary: str
    scope: KomodorScope
    severity: str = "information"
    details: dict[str, str] | None = None


def first_value(values: dict[str, Any], *keys: str) -> str | None:
    for key in keys:
        value = values.get(key)
        if value:
            return str(value)

    return None


def optional_list(value: str | None) -> list[str] | None:
    if value is None or value.strip() == "":
        return None

    return [value]


def map_severity(labels: dict[str, Any], status: str) -> str:
    if status == "resolved":
        return "information"

    severity = str(labels.get("severity", "")).lower()

    if severity in {"critical", "error", "page"}:
        return "error"

    if severity in {"warning", "warn"}:
        return "warning"

    return "information"


def build_event(alert: dict[str, Any]) -> KomodorEvent | None:
    labels = alert.get("labels", {})
    annotations = alert.get("annotations", {})
    status = str(alert.get("status", "firing"))

    cluster = first_value(labels, "cluster")
    if not cluster:
        return None

    alert_name = str(labels.get("alertname", "AlertmanagerAlert"))
    event_type = alert_name[:30]

    summary = (
        annotations.get("summary")
        or annotations.get("description")
        or f"{status.title()}: {alert_name}"
    )

    namespace = first_value(labels, "namespace", "kubernetes_namespace")
    service = first_value(
        labels,
        "service",
        "service_name",
        "app",
        "app_kubernetes_io_name",
        "deployment",
        "statefulset",
        "daemonset",
    )

    scope = KomodorScope(
        clusters=[cluster],
        namespaces=optional_list(namespace),
        servicesNames=optional_list(service),
    )

    details = {
        "status": status,
        "startsAt": str(alert.get("startsAt", "")),
        "endsAt": str(alert.get("endsAt", "")),
        "generatorURL": str(alert.get("generatorURL", "")),
        "fingerprint": str(alert.get("fingerprint", "")),
        **{f"{key}": str(value) for key, value in labels.items()},
    }
    # We want annotations to be included, but we want labels to take precedence in case of key conflicts.
    # so we add annotations after labels with a prefix (if needed).
    [details.update({f"annotation_{key}" if key in details else key: str(value)}) for key, value in annotations.items()]

    return KomodorEvent(
        eventType=event_type,
        summary=str(summary),
        scope=scope,
        severity=map_severity(labels, status),
        details=details,
    )


async def send_to_komodor(event: KomodorEvent) -> None:
    if not KOMODOR_API_KEY:
        raise HTTPException(
            status_code=500,
            detail="KOMODOR_API_KEY environment variable is not set",
        )

    headers = {
        "x-api-key": KOMODOR_API_KEY,
        "content-type": "application/json",
    }

    async with httpx.AsyncClient(timeout=REQUEST_TIMEOUT_SECONDS) as client:
        response = await client.post(
            KOMODOR_API_URL,
            headers=headers,
            json=event.model_dump(exclude_none=True),
        )

    if response.status_code != 201:
        raise HTTPException(
            status_code=502,
            detail={
                "message": "Komodor API request failed",
                "status_code": response.status_code,
                "body": response.text,
            },
        )


@app.get("/healthz")
async def healthz() -> dict[str, str]:
    return {"status": "ok"}


@app.post("/alertmanager")
async def alertmanager_webhook(request: Request) -> dict[str, int]:
    payload = await request.json()
    alerts = payload.get("alerts", [])

    sent = 0
    skipped = 0

    for alert in alerts:
        event = build_event(alert)
        if event is None:
            skipped += 1
            continue

        await send_to_komodor(event)
        sent += 1

    return {"received": len(alerts), "sent": sent, "skipped": skipped}
